pkg/disttask/framework/scheduler/nodes.go

```diff
@@ -116,50 +117,82 @@ func (nm *NodeManager) maintainLiveNodes(ctx context.Context, taskMgr TaskManage
 	nm.prevLiveNodes = currLiveNodes
 }
 
-func (nm *NodeManager) refreshManagedNodesLoop(ctx context.Context, taskMgr TaskManager, slotMgr *SlotManager) {
+func (nm *NodeManager) refreshNodesLoop(ctx context.Context, taskMgr TaskManager, slotMgr *SlotManager) {
 	ticker := time.NewTicker(nodesCheckInterval)
 	defer ticker.Stop()
 	for {
 		select {
 		case <-ctx.Done():
 			return
 		case <-ticker.C:
-			nm.refreshManagedNodes(ctx, taskMgr, slotMgr)
+			nm.refreshNodes(ctx, taskMgr, slotMgr)
 		}
 	}
 }
 
 // TestRefreshedChan is used to sync the test.
 var TestRefreshedChan = make(chan struct{})
 
-// refreshManagedNodes maintains the nodes managed by the framework.
-func (nm *NodeManager) refreshManagedNodes(ctx context.Context, taskMgr TaskManager, slotMgr *SlotManager) {
-	newNodes, err := taskMgr.GetManagedNodes(ctx)
+// refreshNodes maintains the nodes managed by the framework.
+func (nm *NodeManager) refreshNodes(ctx context.Context, taskMgr TaskManager, slotMgr *SlotManager) {
+	newNodes, err := taskMgr.GetAllNodes(ctx)
 	if err != nil {
 		nm.logger.Warn("get managed nodes met error", llog.ShortError(err))
 		return
 	}
-	nodeIDs := make([]string, 0, len(newNodes))
+
 	var cpuCount int
 	for _, node := range newNodes {
-		nodeIDs = append(nodeIDs, node.ID)
 		if node.CPUCount > 0 {
 			cpuCount = node.CPUCount
 		}
 	}
 	slotMgr.updateCapacity(cpuCount)
-	nm.managedNodes.Store(&nodeIDs)
+	nm.nodes.Store(&newNodes)
 
 	failpoint.Inject("syncRefresh", func() {
 		TestRefreshedChan <- struct{}{}
 	})
 }
 
-// GetManagedNodes returns the nodes managed by the framework.
-// return a copy of the managed nodes.
-func (nm *NodeManager) getManagedNodes() []string {
-	nodes := *nm.managedNodes.Load()
-	res := make([]string, len(nodes))
+// GetNodes returns the nodes managed by the framework.
+// return a copy of the nodes.
+func (nm *NodeManager) getNodes() []proto.ManagedNode {
+	nodes := *nm.nodes.Load()
+	res := make([]proto.ManagedNode, len(nodes))
 	copy(res, nodes)
 	return res
 }
+
+func filterByScope(nodes []proto.ManagedNode, targetScope string) []string {
+	var nodeIDs []string
+	haveOther := false
+	haveBackground := false
+	for _, node := range nodes {
+		if node.Role != "" && node.Role != "background" {
+			haveOther = true
+		}
+		if node.Role == "background" {
+			haveBackground = true
+		}
+	}
+
+	for _, node := range nodes {
+		if haveOther {
+			if node.Role == targetScope {
+				nodeIDs = append(nodeIDs, node.ID)
+			}
```

Comment:
if `haveOther=true` and `targetScope=""`, we will always enter this branch

Reply:
nodes with  node1="", node2="", node3="xxx", targetScope="",  filterByScope return node1, node2

Reply:
for this case, it should return `3`
```
nodes:         []proto.ManagedNode{mockManagedNode("1", "1"), mockManagedNode("2", ""), mockManagedNode("3", "background")},
targetScope:   "",
```